facebook/delphiFacebook/R/contingency_tables.R

```diff
@@ -0,0 +1,1068 @@
+#' @param signal_name    name of the signal; used for naming the output file
+#'
+#' @importFrom readr write_csv
+#' @importFrom dplyr arrange_at
+#' 
+#' @export
+write_contingency_tables <- function(data, params, geo_level, groupby_vars)
+{
+  if (!is.null(data) && nrow(data) != 0) {
+    data <- arrange_at(data, groupby_vars)
+  } else {
+    msg_plain(sprintf(
+      "no aggregations produced for grouping variables %s (%s); CSV will not be saved", 
+      paste(groupby_vars, collapse=", "), geo_level
+    ))
+    return()
+  }
+  
+  file_out <- file.path(
+    params$export_dir, sprintf("%s_%s_%s.csv", format(params$start_date, "%Y%m%d"),
+                               geo_level, paste(groupby_vars, collapse="_"))
+  )
+  
+  create_dir_not_exist(params$export_dir)
+  
+  msg_df(sprintf(
+    "saving contingency table data to %-35s",
+    sprintf("%s_%s_%s", format(params$start_date, "%Y%m%d"),
+            geo_level, paste(groupby_vars, collapse="_"))
+  ), data)
+  write_csv(data, file_out)
+}
+
+
+
+#' Get the date of the first day of the previous month
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return Date
+#' 
+#' @importFrom lubridate floor_date
+#' 
+#' @export
+start_of_prev_full_month <- function(date) {
+  return(floor_date(date, "month") - months(1))
+}
+
+#' Get the date of the last day of the previous month
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return Date
+#' 
+#' @importFrom lubridate ceiling_date days
+#' 
+#' @export
+end_of_prev_full_month <- function(date) {
+  if (ceiling_date(date, "month") == date) {
+    return(date)
+  }
+  
+  return(floor_date(date, "month") - days(1))
+}
+
+
+#' Get the date range specifying the previous month
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return list of two Dates
+#' 
+#' @export
+get_range_prev_full_month <- function(date = Sys.Date()) {
+  eom = end_of_prev_full_month(date)
+  
+  if (eom == date) {
+    som = start_of_prev_full_month(date + months(1))
+  } else {
+    som = start_of_prev_full_month(date)
+  }
+  
+  return(list(som, eom))
+}
+
+
+#' Get the date of the first day of the previous week
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return Date
+#' 
+#' @importFrom lubridate floor_date weeks
+#' 
+#' @export
+start_of_prev_full_week <- function(date) {
+  return(floor_date(date, "week") - weeks(1))
+}
+
+
+#' Get the date of the last day of the previous week
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return Date
+#' 
+#' @importFrom lubridate ceiling_date days
+#' 
+#' @export
+end_of_prev_full_week <- function(date) {
+  if (ceiling_date(date, "week") == date) {
+    return(date)
+  }
+  
+  return(floor_date(date, "week") - days(1))
+}
+
+
+#### TODO: should be epiweeks eventually. Already exists a package to calculate?
```

Comment:
The MMWRweek package works but is clunky; lubridate [supports epiweeks](https://lubridate.tidyverse.org/reference/week.html) too